### ISSUE TITLE
Writing plugins in OOP and decorators

### ISSUE THREAD

**Issue by staltz** [ssbc/secret-stack#41]
**TL;DR I want to use classes and decorators to write plugins, and secret-stack basically already supports that, but it needs some small non-breaking tweaks to the implementation.**

-----

In rewriting the gossip plugin, I noticed some hacks. `init()` has to return the object that is used for muxrpc, and some of the methods in gossip's returned object call other methods. So what the gossip plugin did was create a variable that's used in the closure.

```js
{
  name: 'gossip',
  manifest: { /* ... */ },
  // etc
  init: function (ssb, config) {
    // Some initialization here...
    // ...
    
    var gossip = {
      add: function(addr, source) {
        // ...
      },
      connect: function(addr) {
	gossip.add(addr, 'manual')  // <-- THIS HERE
	// ...
      },
    }
    
    // More initialization!
    // ...
    
    return gossip;
  }
}
```

This is a quite convoluted way of doing OOP. I know you can use closures instead of classes and vice-versa, so that's why I began looking into rewriting one of my plugins with classes. The nice thing is that with TypeScript, using `static` and decorators, you can get quite far. Here's a comparison, before and after, based on the EBT plugin:

**Closures and objects**:

```js
var plugin = {
  name: 'ebt',
  version: '1.0.0',
  manifest: {
    replicate: 'duplex',
    request: 'sync',
    block: 'sync',
    peerStatus: 'sync'
  },
  permissions: {
    anonymous: {allow: ['replicate']}
  },
  init: function (sbot, config) {
    // INITIALIZATION CODE 
  
    return {
      replicate: function (opts) { /* ... */ },
      request: function (id, other) { /* ... */ },
      block: function (id) { /* ... */ },
      peerStatus: function (id) { /* ... */ },
    }
  }
}
```

**Classes and decorators in TypeScript:**

```typescript
@plugin('1.0.0')
class ebt {
  constructor(sbot, config) {
    // INITIALIZATION CODE 
  }
  
  @muxrpc('duplex', {anonymous: 'allow'})
  replicate(opts) { /* ... */ }
  
  @muxrpc('sync')
  request(id, other) { /* ... */ }
  
  @muxrpc('sync')
  block(id) { /* ... */ }
  
  @muxrpc('sync')
  peerStatus(id) { /* ... */ }
}
```

You can see the benefit of using classes and decorators here: every method (private or public) can call each other through the `this` keyword without doing hoops with closures, and you can co-locate the manifest and permissions of a method at the definition of that method.

So, I'm not trying to sell you'all on OOP (because writing plugins in plain object style should still be supported), but here's my point:

I managed to implement the above rather easily. The `@plugin` decorator attaches the static properties `.version` and `.init` to the class, and the `@muxrpc` decorators mutate the static `manifest` and `permissions` properties. So all the fields that secret-stack and muxrpc expect are there, everything is fine.

**But:** some implementation details inside secret-stack assume that we're giving it an object, not a class (which is in runtime a JS function), for instance: https://github.com/ssbc/secret-stack/blob/195151ad97b0c5e8954e72f35c4cca193fb47e60/api.js#L56

This line will receive the class and *think* that it's a function, and then wrap it with an `init`. That sort of breaks the support for this idea of mine. Now, of course, we could just PR & patch that file so that the check for `function` happens later (in the sense that if the argument we received has `.manifest`, `.version`, `.permissions`, don't even check that it's a function), BUT there are even more weird assumptions (weird because I don't understand them all) inside secret-stack that make it hard to actually pass in these classes as plugins.

I believe that support for classes as plugins is achievable, but before I spend more time into this (which is somewhat cosmetic compared to the bigger picture), I'd like to know if a PR would be accepted, as well as open feedback on this idea.

**Comment by cblgh** [comment]
i'm not invested in this, but i can imagine a few questions that would come up:
* how much work is it to get setup with writing typescript on a computer?
* are you free to write non-typescript code in the plugin classes themselves? (i'm pretty sure you are, but good to mention)

**Comment by staltz** [comment]
@cblgh setting up typescript is like setting up a linter or `standard`. First, npm install (as devDependency maybe) `typescript`, then create a config file `tsconfig.json` using `tsc --init` (`tsc` is the binary installed by typescript). To compile, run `tsc`. Usually, we put the compiled javascript in a folder such as `lib`, while the typescript is in `src`, but all this is configurable.

I *think* these decorators would also work with Babel: https://babeljs.io/docs/en/babel-plugin-proposal-decorators . Decorators are a stage 2 proposal: https://github.com/tc39/proposal-decorators

Second question: yes, javascript code should run totally fine. Overall this issue does not (should not) affect javascript users in noticeable any way. A plugin written in a typescript class could still be used in javascript, and other plugins authored the classic way would still work.

**Comment by dominictarr** [comment]
the whole supply plugin as a function was a SLAGIATT but only got used in https://github.com/ssbc/ssb-master (and probably some tests)

So if you replaced updated those, you could change the behaviour to calling the function, expecting it to return a object...

I'm still gonna write it the JS style, though. I've written code in class based languages, I just don't feel it expresses the way I want to solve problems I guess? anyway, if it calls it with new, and my code just returns an object, that's fine, right?

**Comment by dominictarr** [comment]
of course the example in the decorators repo is logging, that's always the example in things like this!

The problem is it's hard to come up with other examples. We already have this pattern, using [hoox](https://www.npmjs.com/package/hoox) but it makes for really confusing indirection. You need a birds eye view (i.e. you wrote all the code) to know that, ah but this function is decorated by some other place in a different repo when you install another plugin... there are a few places this happens in ssb plugins currently, mainly around authentication. Other places have moved towards a more explicit passing of handlers (for example adding transports).

Depject also used a similar pattern - but then most of the people using it on the frontend refactored everything to reduce indirection, which also ruined the thing that indirection is good for (I think we should just drop depject entirely, but legacy code...)

decorators or hooks doesn't automatically give you a modular system, it just lets you incoherently tangle code together. To get a good module system, you need to think very hard about how portions of code interface with each other, and then make a tasteful design.

**Comment by staltz** [comment]
## Comments about this idea (_plugin as a class_)

Hey @dominictarr, thanks for giving this a thought. I think there are two important comments:

> I've written code in class based languages, I just don't feel it expresses the way I want to solve problems I guess?

_Plugin as a class_ would not change existing code or compel people to start writing in this style, this is only about tweaking the implementation to add one more supported way of writing plugins, for people like me.

> decorators

Note that I'm not advocating for decorators that "decorate" third-party code in the way that `hoox` does. These decorators would only be used in "first-party" code, i.e. just to replace `manifest`. This would look rather similar to `muxrpc-validation` which "decorates" rpc methods in "first-party" style:

```diff
 // in the gossip plugin, muxrpc-validation just 'enhances' a method
-add: function (addr, source) {
+add: valid.sync(function (addr, source) {
```

```diff
 class gossip {
   // ...

   // in this gossip plugin as a class, the decorator just 
   // 'enhances' the plugin by adding a field to the manifest, `enable: 'sync'`
+  @muxrpc('sync')
   enable(type) {
     // ... 
   }
  
 }
```

So my proposal for decorators is very different to depject or hoox or any type of aspect oriented programming, because the decorators are used strictly locally.

## Implementation details to get it working in secret-stack

### Conflict with plugin as a function

> the whole supply plugin as a function was a SLAGIATT but only got used in https://github.com/ssbc/ssb-master (and probably some tests)
> 
> So if you replaced updated those, you could change the behaviour to calling the function, expecting it to return a object...

I found a rather easy way of supporting both (plugin as a function, and plugin as a class) in a non-breaking way. It's just about shuffling some if conditions around to be in the right order. This is easy to fix.

### secret-stack attaching hoox to everything

My biggest challenge actually is that my plugin as a class has private methods, but there are no such thing as private methods in javascript classes, and secret-stack navigates the whole returned `api` object, attaching hooks to each thing. In my case this turns out to cause a recursive infinite loop, because one of my private fields is `sbot`, which contains `sbot.gossip`, which contains `sbot.gossip.sbot`, which contains `sbot.gossip.sbot.gossip`, etc.

Currently I'm trying to find a way of forcing private fields or making those fields non-enumerable so that secret-stack would ignore it when iterating over the fields of the `api` object. I'll let you know if I find a solution. But one easy solution would be a naming convention: any field prefixed with `_` is considered private and secret-stack will ignore it when attaching hoox to everything else.

**Comment by staltz** [comment]
Nevermind, I found a way to support this! The only change that is needed in secret-stack is PR #42. I intend to make a new package `secret-stack-decorators`, and only if there is enough interest, we could consider moving that into `secret-stack/decorators`.

**Comment by staltz** [comment]
Alright, published this https://gitlab.com/staltz/secret-stack-decorators
Now if we get PR #42 merged and released, then I can leave y'all alone. : )
